### PR TITLE
[RAG] Add Ray implementation for distributed retrieval

### PR DESCRIPTION
**Pull request by amogkam**
# What does this PR do?
This PR adds a new distributed retriever implementation for RAG built on Ray, as an alternative to the current retriever implementation that uses torch.distributed. With Ray it's possible to load the index on multiple processes instead of just the rank 0 training worker, allowing fine tuning to scale out better to multiple GPUs, and also allowing the index to potentially be fit in GPU memory. This also removes a core dependency on Pytorch, allowing a Tensorflow implementation of `finetune.py`.

This PR also makes changes to support finetune.py with Pytorch Lightning >v1.0.

A benchmark of Pytorch distribtued retrieval vs. Ray distributed retrieval
image

## Implementation Details
In the current Pytorch retrieval implementation, the index is loaded once on just the rank 0 training workers. Training worker 0 gathers the inputs from all other workers, performs the index lookup, and scatters the results back to the other workers.
image

With the Ray implementation, the index is loaded on *separate* processes, which are referred to as Ray actors. Each training worker randomly selects a retrieval actor to query for documents and Ray handles all the communication between the processes. Because the index can be loaded in *multiple* processes, training can scale up since no synchronization needs to happen for the index lookup.
image

Note that Pytorch Lightning is still handling distributed *training*, but Ray manages distributed *retrieval*. Because PTL calls the entire training script under the hood multiple times, we have to use Ray's named actors feature (https://docs.ray.io/en/master/actors.html?highlight=named%20actors#named-actors) allowing the retrieval actors to be referenced by all training processes. The use of named actors is necessitated by how PTL handles distributed training, and a simpler approach could probably be used for a Tensorflow implentation.

## Testing Strategy
Unit tests were added to `test_distributed_retriever.py`. Note that the local Ray cluster for the tests had to be started with `local_mode=True` because the test file modifies `sys.path` and these changes are not propagated to remote processes. See https://stackoverflow.com/questions/54338013/parallel-import-a-python-file-from-sibling-folder for more info.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by lhoestq** [comment]
Hi ! This looks awesome :)
I was about to create a PR that fixes the init_ddp_connection in finetune.py and that adds a test script to make sure the finetuning script works as expected. With minimal changes on my side I can easily reduce conflicts between our two changes to finetune.py (I guess I'll just reuse the CustomAccelerator). Does that sound good to you ?

**Comment by amogkam** [comment]
@lhoestq yes that sounds great!

**Comment by shamanez** [comment]
Hi seems like finetune.sh is not working in multi gpu training.

**Comment by shamanez** [comment]
@amogkam 


Hi seems like finetune.sh is not working in multi gpu training.

**Comment by amogkam** [comment]
@shamanez Hmm that's odd, I was able to get this working on a single node with 4 GPUs. Do you have a stack trace?

**Comment by shamanez** [comment]
> @shamanez Hmm that's odd, I was able to get this working on a single node with 4 GPUs. Do you have a stack trace?

I tried to run without RAY , but with pytorch DDP. Here is the error I got.


Selection_006
Selection_007


This is the command-line argument I used, 
`python examples/rag/finetune.py     --data_dir ./examples/rag/test_data/dummy_seq2seq     --output_dir ./examples/rag/outputs   --model_name_or_path facebook/rag-token-base      --model_type rag_sequence       --do_train     --do_predict     --n_val -1     --val_check_interval 0.25     --train_batch_size 1     --eval_batch_size 1     --max_source_length 128     --max_target_length 25     --val_max_target_length 25     --test_max_target_length 25     --label_smoothing 0.1     --dropout 0.1     --attention_dropout 0.1     --weight_decay 0.001     --adam_epsilon 1e-08     --max_grad_norm 0.1     --lr_scheduler polynomial     --learning_rate 3e-05     --num_train_epochs 100     --warmup_steps 500     --gradient_accumulation_steps 1    --index_name custom    --passages_path ./examples/rag/data/my_knowledge_dataset    --index_path  ./examples/rag/data/my_knowledge_dataset_hnsw_index.faiss --gpus 2
`

**Comment by amogkam** [comment]
I think this has to do with using a custom index which I didn't try out. Can you try with just the wiki_dpr index to confirm? It seems like the training workers are expecting a tensor of type float, but a tensor of type double is being sent instead. I think the fix might just be to set an explicit target_type in line 137 of distributed_pytorch_retriever.py- @lhoestq does this seem right?

**Comment by shamanez** [comment]
Ok I will also give it a try

**Comment by amogkam** [comment]
@lhoestq now that https://github.com/huggingface/transformers/pull/8585 is merged, should I mark this PR as ready for review?

**Comment by lhoestq** [comment]
Yes indeed ! Feel free to set this PR to ready for review

Also it looks like the CI fails because of a failed import of `ray`.
To fix that you need to move the import of ray into the test functions decorated with `require_distributed_retrieval `.

You should also add `ray` to the test dependencies, or the test will simply be ignored

**Comment by amogkam** [comment]
@lhoestq CI is passing now!

**Comment by amogkam** [comment]
@lhoestq any ETA on when this PR can get reviewed? Thanks

**Comment by lhoestq** [comment]
Hi ! I've already started to look at the changes and it looks pretty good so far :) I'll finish my review soon, probably tomorrow

**Comment by amogkam** [comment]
Awesome thanks!

File: examples/rag/README.md [hunk 1]
```diff
@@ -47,6 +47,39 @@ python examples/rag/consolidate_rag_checkpoint.py \
...
+
+python examples/rag/finetune.py \
+    --data_dir $DATA_DIR \
+    --output_dir $OUTPUT_DIR \
+    --model_name_or_path $MODEL_NAME_OR_PATH \
+    --model_type rag_sequence \
+    --fp16 \
+    --gpus 8
+    --distributed_retriever ray \
+    --num_retrieval_workers 4
```
Comment (lhoestq) [inline examples/rag/README.md:77]: maybe add an example for torch as well ?

File: examples/rag/finetune_rag.py [hunk 2]
```diff
@@ -442,6 +466,20 @@ def add_retriever_specific_args(parser):
...
+            default="pytorch",
+            help="What implementation to use for distributed retriever? If "
+            "pytorch is selected, the index is loaded on training "
+            "worker 0, and torch.distributed is used to handle "
+            "communication between training worker 0, and the other "
+            "training workers. If ray is selected, the Ray library is "
+            "used to create load the index on separate processes, "
+            "and Ray handles the communication between the training "
+            "workers and the retrieval actors.",
+        )
```
Comment (lhoestq) [inline examples/rag/finetune_rag.py:482]: Nice !

File: examples/rag/finetune_rag.py [hunk 3]
```diff
@@ -509,4 +599,29 @@ def main(args=None, model=None) -> GenerativeQAModule:
...
+
+    # Pytorch Lightning Profiler
+    parser.add_argument(
+        "--profile",
+        action="store_true",
+        help="If True, use pytorch_lightning.profiler.AdvancedProfiler to " "profile the Trainer.",
+    )
+
+    # Ray cluster address
+    parser.add_argument(
```
Comment (lhoestq) [inline examples/rag/finetune_rag.py:615]: Maybe group the ray-specific args (`ray-address` and `num_retrieval_workers `) with a function `add_ray_specific_args` ?

File: examples/rag/test_distributed_retriever.py [hunk 4]
```diff
@@ -175,50 +214,111 @@ def get_dummy_custom_hf_index_retriever(self, init_retrieval: bool, from_disk: b
...
+            self.assertEqual(retrieved_doc_embeds.shape, (2, n_docs, self.retrieval_vector_size))
+            self.assertEqual(len(doc_dicts), 2)
+            self.assertEqual(sorted(doc_dicts[0]), ["embeddings", "id", "text", "title"])
+            self.assertEqual(len(doc_dicts[0]["id"]), n_docs)
+            self.assertEqual(doc_dicts[0]["id"][0], "1")  # max inner product is reached with second doc
+            self.assertEqual(doc_dicts[1]["id"][0], "0")  # max inner product is reached with first doc
+            self.assertListEqual(doc_ids.tolist(), [[1], [0]])
+
+        test_retriever(self.get_dummy_pytorch_distributed_retriever(init_retrieval=True))
+        test_retriever(self.get_dummy_ray_distributed_retriever(init_retrieval=True))
```
Comment (lhoestq) [inline examples/rag/test_distributed_retriever.py:279]: Could you separate the pytorch test and the ray test in different functions ?
And add a `require_ray` only for the one that uses ray ?

File: examples/rag/test_finetune_rag.py [hunk 5]
```diff
@@ -87,10 +88,12 @@ def _run_finetune(self, gpus: int):
 
     @require_torch_gpu
     def test_finetune_gpu(self):
-        result = self._run_finetune(gpus=1)
-        self.assertGreaterEqual(result["test"][0]["test_avg_em"], 0.2)
+        for distributed_retriever in ["pytorch", "ray"]:
+            result = self._run_finetune(gpus=1, distributed_retriever=distributed_retriever)
+            self.assertGreaterEqual(result["test"][0]["test_avg_em"], 0.2)
```
Comment (lhoestq) [inline examples/rag/test_finetune_rag.py:93]: same here

File: src/transformers/integrations.py [hunk 6]
```diff
@@ -47,8 +47,16 @@
     import ray  # noqa: F401
 
     _has_ray = True
+    try:
```
Comment (lhoestq) [inline src/transformers/integrations.py:50]: adding ray integration here cc @LysandreJik

**Review by lhoestq — COMMENTED** [review]
Really good ! thank you for adding ray support for RAG fine-tuning :)
And the speed up comparing to using only one worker for retrieval is pretty cool.

I left a few comments, mainly about separating the pytorch tests from the ray tests.

**Comment by LysandreJik** [comment]
@sgugger it would be cool if you could review as this changes some things in the trainer/integrations.

**Review by sgugger — CHANGES_REQUESTED** [review]
There are other instance of `is_ray_available` to change to `is_ray_tune_available` if we go with the name change:
- in `integrations.py`, inside the function `hp_params` and `default_hp_search_backend`
- in `trainer_utils.py`, inside the function `default_hp_space_ray`

The main `__init__` should also be updated to provide the two functions.

File: examples/rag/test_finetune_rag.py (hunk 5 above)
Comment (amogkam) [reply examples/rag/test_finetune_rag.py]: Done!

File: examples/rag/test_distributed_retriever.py (hunk 4 above)
Comment (amogkam) [reply examples/rag/test_distributed_retriever.py]: Done!

File: examples/rag/finetune_rag.py (hunk 3 above)
Comment (amogkam) [reply examples/rag/finetune_rag.py]: Done!

File: examples/rag/README.md (hunk 1 above)
Comment (amogkam) [reply examples/rag/README.md]: Currently distributed_retriever defaults to pytorch so an example command for this would just be the same as the command earlier in the Readme. I added a sentence saying that the default is pytorch though.

**Comment by amogkam** [comment]
Hi @lhoestq @sgugger I addressed the feedback you guys gave. Do you think you can take another look? Thanks

**Comment by sgugger** [comment]
Hi there, sorry for the delay. Could you close and reopen your PR? Because of a bad force-push on our side, the diff has become unreadable. Also, the examples folder has slightly changed structure, so you might need to move the folder.

Ping me, @patrickvonplaten  and @LysandreJik on the PR you reopen and we'll look at it quickly.

**Comment by amogkam** [comment]
Opened a new one here: https://github.com/huggingface/transformers/pull/9197!
